pyScienceMode2/rehastim_interface.py

```diff
@@ -371,3 +367,527 @@ def get_motomed_torque(self) -> float:
+
+    def ll_stop(self):
+        """
+        Stop the device lower level.
+        """
+        packet_number = self.get_next_packet_number()
+        if not sciencemode.lib.smpt_send_ll_stop(self.device, packet_number):
+            raise RuntimeError("Low level stop failed.")
+        self.log(
+            "Low level stopped",
+            "Command sent to rehastim: {}".format(self.TypeRehap24(sciencemode.lib.Smpt_Cmd_Ll_Stop).name),
+        )
+        self._get_last_ack()
+
+    def init_stimulation(self, list_channels: list, stop_all_on_error: bool = True):
+        """
+        Initialize the mid level stimulation on the device. It is used for defining a stimulation pattern
+
+        Parameters
+        ----------
+        list_channels : list
+            Channels to stimulate.
+        stop_all_on_error : bool
+            If flag is set to True ,stop stimulation if one channel has an error.
+        """
+        if self.stimulation_started:
+            self.stop_stimulation()
+        self.list_channels = list_channels
+        check_unique_channel(list_channels)
+        self.electrode_number = calc_electrode_number(self.list_channels)
+
+        ml_init = sciencemode.ffi.new("Smpt_ml_init*")
+        ml_init.stop_all_channels_on_error = stop_all_on_error
+        ml_init.packet_number = self.get_next_packet_number()
+
+        if not sciencemode.lib.smpt_send_ml_init(self.device, ml_init):
+            raise RuntimeError("Failed to start stimulation")
+        self.log(
+            "Stimulation initialized",
+            "Command sent to rehastim: {}".format(self.TypeRehap24(sciencemode.lib.Smpt_Cmd_Ml_Init).name),
+        )
+        self._get_last_ack()
+
+    def start_stimulation(self, upd_list_channels: list, stimulation_duration: int | float = None, safety: bool = True):
+        """
+        Start the mid level stimulation on the device.
+
+        Parameters
+        ----------
+        stimulation_duration : int | float
+            Duration of the stimulation in seconds.
+        upd_list_channels : list
+            Channels to stimulate.
+        safety : bool
+            Set to True if you want to check the pulse symmetry. False otherwise.
+        """
+        if not stimulation_duration:
+            raise ValueError("Please indicate the stimulation duration")
+        if upd_list_channels is not None:
+            new_electrode_number = calc_electrode_number(upd_list_channels)
+            if new_electrode_number != self.electrode_number:
+                raise RuntimeError("Error update: all channels have not been initialised")
+        self.list_channels = upd_list_channels
+
+        ml_update = sciencemode.ffi.new("Smpt_ml_update*")
+        ml_update.packet_number = self.get_next_packet_number()
+
+        #  Check if points are provided for each channel stimulated
+        for channel in upd_list_channels:
+            if not channel.is_pulse_symmetric(safety=safety):
```

Comment:
Why there is a safety parameter to this function as it is supposed to be called only if safety mode. I would put a statement if safety: is pulse symetric.

Reply:
Fixed.